// angular-star-rating/esm2020/src/angular-star-rating.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { StarRatingComponent } from './components/star-rating/star-rating.component';
import { StarRatingControlComponent } from './components/star-rating-control/star-rating-control.component';
import { StarRatingConfigService } from './services/star-rating-config.service';
import * as i0 from "@angular/core";
const DECLARATIONS = [StarRatingComponent, StarRatingControlComponent];
const EXPORTS = [DECLARATIONS];
export class StarRatingModule {
    static forRoot() {
        return {
            ngModule: StarRatingModule,
            providers: [StarRatingConfigService],
        };
    }
    static forChild() {
        return {
            ngModule: StarRatingModule,
            providers: [],
        };
    }
}
StarRatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: StarRatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StarRatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: StarRatingModule, declarations: [StarRatingComponent, StarRatingControlComponent], imports: [CommonModule], exports: [StarRatingComponent, StarRatingControlComponent] });
StarRatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: StarRatingModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: StarRatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [DECLARATIONS],
                    exports: [EXPORTS],
                }]
        }] });
